File: dfs/coins_collector/coins_collector.py

class CoinsCollector:  # LeetCode Q.2603.
    def __init__(self):
        self.graph: list[set[int]] = []
        self.leaves: list[bool] = []
        self.leaves_parents: list[bool] = []

        self.empty_parents: set[int] = set()
        self.max_empty_parents_dist = 1

        self.required_visits: list[bool] = []
        self.visited: list[bool] = []
        self.must_visit_edges = 0

    def _dfs_required_visit_nodes(self, node: int) -> None:
        self.visited[node] = True
        for neighbor in self.graph[node]:
            if self.leaves_parents[neighbor] and not self.visited[neighbor]:
                self.required_visits[node] = True
                self._dfs_required_visit_nodes(neighbor)

    def _dfs_must_visit_edges(self, parent: int, node: int) -> bool:
        self.visited[node] = True
        for neighbor in self.graph[node]:
            if not self.visited[neighbor]:
                if self._dfs_must_visit_edges(node, neighbor):
                    self.required_visits[node] = True

        if self.required_visits[node] and parent != -1:
            self.must_visit_edges += 1
        return self.required_visits[node]

    def _dfs_max_empty_parents_dist(self, node: int) -> int:
        self.visited[node] = True
        children_dists: list[int] = []
        for neighbor in self.graph[node]:
            if neighbor in self.empty_parents:
                if not self.visited[neighbor]:
                    children_dists.append(self._dfs_max_empty_parents_dist(neighbor))

        local_max_dist = 1
        if not children_dists:
            return local_max_dist

        children_dists.sort()
        local_max_dist = children_dists[-1]

        if len(children_dists) >= 2:
            local_max_dist += children_dists[-2]

        if local_max_dist > self.max_empty_parents_dist:
            self.max_empty_parents_dist = local_max_dist

        return children_dists[-1] + 1

    def collect_coins(self, coins: list[int], edges: list[list[int]]) -> int:
        self.graph.clear()
        self.graph.extend([set() for _ in range(len(coins))])
        for node_1, node_2 in edges:  # Undirected edges.
            self.graph[node_1].add(node_2)
            self.graph[node_2].add(node_1)

        for node, coin in enumerate(coins):
            while coin == 0 and len(self.graph[node]) == 1:
                # Graph reduction: keep removing leaves w/o coins.
                neighbor = list(self.graph[node])[0]
                self.graph[neighbor].remove(node)
                self.graph[node].clear()
                node, coin = neighbor, coins[neighbor]

        self.leaves.clear()
        self.leaves.extend([False] * len(coins))
        self.leaves_parents.clear()
        self.leaves_parents.extend([False] * len(coins))
        for node in range(len(coins)):
            if coins[node] == 1 and len(self.graph[node]) == 1:
                self.leaves[node] = True

                neighbor = list(self.graph[node])[0]
                if coins[neighbor] == 0 or len(self.graph[neighbor]) > 1:
                    self.leaves_parents[neighbor] = True

        self.empty_parents.clear()
        self.max_empty_parents_dist = 1

        self.required_visits.clear()
        self.required_visits.extend([False] * len(coins))
        for node in range(len(coins)):
            if self.leaves_parents[node]:
                is_empty_parent = True

                for neighbor in self.graph[node]:
                    if not self.leaves_parents[neighbor] and not self.leaves[neighbor]:
                        self.required_visits[neighbor] = True
                        is_empty_parent = False

                if is_empty_parent:
                    self.empty_parents.add(node)

        if True not in self.required_visits:
            if not self.empty_parents:
                return 0

            self.visited.clear()
            self.visited.extend([False] * len(coins))
            self._dfs_max_empty_parents_dist(list(self.empty_parents)[0])
            return max(0, (self.max_empty_parents_dist - 2) * 2)

        self.visited.clear()
        self.visited.extend([False] * len(coins))
        must_visit_node = -1
        for node in range(len(coins)):
            if self.required_visits[node] and not self.visited[node]:
                must_visit_node = node
                self._dfs_required_visit_nodes(node)

        if must_visit_node != -1:
            self.must_visit_edges = 0
            self.visited.clear()
            self.visited.extend([False] * len(coins))
            self._dfs_must_visit_edges(-1, must_visit_node)

        return self.must_visit_edges * 2

File: dfs/coins_collector/coins_collector.cpp
#include <vector>
#include <unordered_set>
#include <algorithm>
using namespace std;

class CoinsCollector
{ // LeetCode Q.2603.
private:
    vector<unordered_set<int>> graph;

    vector<bool> leaves, leavesParents; // If a node is a coin leave or parent of coin leave.
    vector<bool> requiredVisits;        // If a node must be visited to collect all coins.
    vector<bool> visited;               // For DFS visited marks.

    void dfsRequiredVisitNodes(int node)
    {
        visited[node] = true;
        for (auto neighbor : graph[node])
        {
            if (leavesParents[neighbor] == true && visited[neighbor] == false)
            {
                requiredVisits[node] = true;
                dfsRequiredVisitNodes(neighbor);
            }
        }
    }

    int mustVisitEdges = 0;

    bool dfsMustVisitEdges(int parent, int node)
    {
        visited[node] = true;
        for (auto neighbor : graph[node])
        {
            if (visited[neighbor] == false)
            {
                if (dfsMustVisitEdges(node, neighbor))
                {
                    requiredVisits[node] = true;
                }
            }
        }

        if (requiredVisits[node] && parent != -1)
        {
            mustVisitEdges++;
        }
        return requiredVisits[node];
    }

    unordered_set<int> emptyParents;
    int maxEmptyParentsDist = 1;

    int dfsMaxEmptyParentsDist(int node)
    {
        visited[node] = true;
        vector<int> childrenDists;

        for (auto neighbor : graph[node])
        {
            if (emptyParents.find(neighbor) != emptyParents.end())
            {
                if (visited[neighbor] == false)
                {
                    childrenDists.push_back(dfsMaxEmptyParentsDist(neighbor));
                }
            }
        }

        int localMaxDist = 1;
        if (childrenDists.empty())
        {
            return localMaxDist;
        }

        sort(childrenDists.begin(), childrenDists.end());
        localMaxDist = childrenDists.back();

        if (childrenDists.size() >= 2)
        {
            localMaxDist += childrenDists[childrenDists.size() - 2];
        }
        if (localMaxDist > maxEmptyParentsDist)
        {
            maxEmptyParentsDist = localMaxDist;
        }

        return childrenDists.back() + 1;
    }

public:
    int collectCoins(vector<int> &coins, vector<vector<int>> &edges)
    {
        graph.resize(coins.size());
        for (auto edge : edges)
        { // Undirected edges.
            graph[edge[0]].insert(edge[1]);
            graph[edge[1]].insert(edge[0]);
        }

        // Graph reduction: keep removing leaves w/o coins.
        for (int node = 0; node < coins.size(); node++)
        {
            while (coins[node] == 0 && graph[node].size() == 1)
            {
                int neighbor = *graph[node].begin();
                graph[neighbor].erase(node);
                graph[node].clear();
                node = neighbor;
            }
        }

        leaves.resize(coins.size(), false);
        leavesParents.resize(coins.size(), false);

        for (int node = 0; node < coins.size(); node++)
        {
            if (coins[node] == 1 && graph[node].size() == 1)
            {
                leaves[node] = true;

                int neighbor = *graph[node].begin();
                if (coins[neighbor] == 0 || graph[neighbor].size() > 1)
                {
                    leavesParents[neighbor] = true;
                }
            }
        }

        bool parentsAllEmpty = true;
        requiredVisits.resize(coins.size(), false);

        for (int node = 0; node < coins.size(); node++)
        {
            if (leavesParents[node])
            {
                bool isEmptyParent = true;

                for (auto neighbor : graph[node])
                {
                    if (leavesParents[neighbor] == false && leaves[neighbor] == false)
                    {
                        requiredVisits[neighbor] = true;
                        isEmptyParent = false, parentsAllEmpty = false;
                    }
                }

                if (isEmptyParent)
                {
                    emptyParents.insert(node);
                }
            }
        }

        if (parentsAllEmpty)
        {
            if (emptyParents.empty())
            {
                return 0;
            }
            visited.resize(coins.size(), false);
            dfsMaxEmptyParentsDist(*emptyParents.begin());
            return max(0, (maxEmptyParentsDist - 2) * 2);
        }

        visited.clear(); // Must clear before resize to reset all values to false.
        visited.resize(coins.size(), false);

        int mustVisitNode = -1;

        for (int node = 0; node < coins.size(); node++)
        {
            if (requiredVisits[node] && visited[node] == false)
            {
                mustVisitNode = node;
                dfsRequiredVisitNodes(node);
            }
        }

        if (mustVisitNode != -1)
        {
            visited.clear(); // Must clear before resize to reset all values to false.
            visited.resize(coins.size(), false);
            dfsMustVisitEdges(-1, mustVisitNode);
        }
        return mustVisitEdges * 2;
    }
};